### PR TITLE
[feat]: [CDS-58067]: label sha support for GAR

### PR DESCRIPTION
**Pull request by Abhishek-CDC**
## Describe your changes
Label and SHA to be added to service output section.
validate and construct image path if SHA passed in tag field.
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)

### CODE REVIEW

**Comment by Abhishek-CDC** [comment]
trigger ut3

**Comment by vivekkarnatiharness** [comment]
*[`950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/artifacts/gar/GarArtifactDelegateResponseTest.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/45752#-NRkuk9l1pFW-T7kCePM:-NRkuk9l1pFW-T7kCePN:bafdas6) ([raw file](https://github.com/harness/harness-core/blob/da0501e18e250e691f518cdfb1b984fc3c35c3fe/950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/artifacts/gar/GarArtifactDelegateResponseTest.java#L36)):*
> ```Java
>   public void testDescribe() {
>     ArtifactBuildDetailsNG buildDetailsNG = ArtifactBuildDetailsNG.builder().build();
>     garDelegateResponse = new GarDelegateResponse(buildDetailsNG, ArtifactSourceType.ECR, "v3.0", new HashMap<>());
> ```

Change this,It should be GAR

**Comment by Abhishek-CDC** [comment]
*[`950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/artifacts/gar/GarArtifactDelegateResponseTest.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/45752#-NRkuk9l1pFW-T7kCePM:-NRl3SGb4muI4YuWzqdd:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/da0501e18e250e691f518cdfb1b984fc3c35c3fe/950-delegate-tasks-beans/src/test/java/io/harness/delegate/task/artifacts/gar/GarArtifactDelegateResponseTest.java#L36)):*
<details><summary><i>Previously, vivekkarnatiharness wrote…</i></summary><blockquote>

Change this,It should be GAR
</blockquote></details>

Done

**Review by vivekkarnatiharness — COMMENTED** [review]
> Reviewed 10 of 11 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: 13 file(s) to review. 1 of 2 approvals obtained.

**Review by vtxorxwitty — DISMISSED** [review]
> Reviewed 6 of 11 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: 12 file(s) to review. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java` line 669 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqwZXX36Yehfx5JU6P:-NRqwZXX36Yehfx5JU6Q:bev150o) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java#L669)):*
> ```Java
>       return Collections.emptyMap();
>     }
>     return EmptyPredicate.isNotEmpty(artifactDelegateResponse.getLabel()) ? artifactDelegateResponse.getLabel()
> ```

nit: can we just say return artifactDelegateResponse.getLabel() 
The above If condition can say ```if(isEmpty(response.getLabel())```

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java` line 61 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqx5Vp1g1S-Rle7ZTU:-NRqx5Vp1g1S-Rle7ZTV:bmp2k9d) ([raw file](https://github.com/harness/harness-core/blob/fe6e409abc7113da6c4446d44fcb164fccc53d84/930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java#L61)):*
> ```Java
>     BuildDetailsInternal lastSuccessfulBuild;
>     GarInternalConfig garInternalConfig;
>     ArtifactMetaInfo artifactMetaInfo;
> ```

is this removal intentional ?

**Comment by yogesh-chauhan** [comment]
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 129 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqxWRr4RVRMOm9SzFH:-NRqxWRr4RVRMOm9SzFI:bj0lvmr) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L129)):*
> ```Java
> 
>     try {
>       ArtifactMetaInfo artifactMetaInfo1 = getArtifactMetaInfoV1(garinternalConfig, version);
> ```

can we use a better name for ```artifactMetaInfo1```

**Comment by yogesh-chauhan** [comment]
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 143 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqxffK9KxlWX4uUGj7:-NRqxffK9KxlWX4uUGj8:blm1un2) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L143)):*
> ```Java
>       }
>     } catch (Exception e) {
>       errorMessage = e.getMessage();
> ```

the same error message object is being overriden here ( from the previous try-catch block )

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 5 of 11 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 10 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: 1 file(s) to review. 0 of 2 approvals obtained.

**Comment by Abhishek-CDC** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java` line 61 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqx5Vp1g1S-Rle7ZTU:-NRr-_G45ELgyzKIKFS-:b-ie1ljp) ([raw file](https://github.com/harness/harness-core/blob/fe6e409abc7113da6c4446d44fcb164fccc53d84/930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java#L61)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

is this removal intentional ?
</blockquote></details>

In previous PR we had added method for getting ArtifactMetaInfo. The verify build method just make an API call with that particular build id and if it is successful it returns required info which doesn't use API response. 
In this PR we are replacing that API to fetch sha and labels (ArtifactMetaInfo) so that we can validate and fetch sha and label details from same call.
ArtifactMetaInfo is added to BuildDetailsInternal in verifyBuild method, So we don't need to call this API separately.

**Comment by Abhishek-CDC** [comment]
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 143 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqxffK9KxlWX4uUGj7:-NRr1JSz3qWc3HdlFe78:bnzh4d9) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L143)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

the same error message object is being overriden here ( from the previous try-catch block )
</blockquote></details>

When image manifest API call fails, getArtifactMetaInfoV2/getArtifactMetaInfoV1 method will throw error.
If any one of them is successful we would not throw exception from this method.
if both are not successful then reason must be same as difference in both call is of 'Accept' header only which specifies accepted media type.

**Comment by Abhishek-CDC** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java` line 669 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqwZXX36Yehfx5JU6P:-NRr3-Lx0loTuI7a0c4G:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java#L669)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

nit: can we just say return artifactDelegateResponse.getLabel() 
The above If condition can say ```if(isEmpty(response.getLabel())```
</blockquote></details>

Done

**Comment by Abhishek-CDC** [comment]
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 129 at r5](https://reviewable.io/reviews/harness/harness-core/45752#-NRqxWRr4RVRMOm9SzFH:-NRr33SV4Cr1p1fJfpFw:byao107) ([raw file](https://github.com/harness/harness-core/blob/704b065ab3c3cbfbbd37d2da844019e59f738e3f/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L129)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we use a better name for ```artifactMetaInfo1```
</blockquote></details>

named after schemaVersion of API response

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 1]
```diff
@@ -117,41 +122,49 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+        artifactMetaInfo.setSha(artifactMetaInfoSchemaVersion1.getSha());
+        artifactMetaInfo.setLabels(artifactMetaInfoSchemaVersion1.getLabels());
       }
-      if (!response.isSuccessful()) {
-        log.error("Request not successful. Reason: {}", response);
-        if (!isSuccessful(response.code(), response.errorBody().toString())) {
-          throw NestedExceptionUtils.hintWithExplanationException("Unable to fetch the versions for the package",
-              "Please check region field", new InvalidArtifactServerException(response.message(), USER));
-        }
+    } catch (Exception e) {
```
Comment (yogesh-chauhan) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:134]: lets log exception at all places

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: 2 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 1 above)
Comment (Abhishek-CDC) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:135]: Done

**Review by puthrayaharness — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 5 of 11 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 10 of 11 files at r4, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: 1 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by Abhishek-CDC** [comment]
trigger messagecheck

**Review by utkarshchoubey-harness — APPROVED** [review]
:lgtm:

> Reviewed 5 of 11 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 10 of 11 files at r4, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752#-:-NS0MD1WA-2t_6ylw2tF:bnfp4nl)* status: 2. unresolved discussions.

**Review by utkarshchoubey-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752#-:-NS0NWuZ15OEPJVOHcJv:bnfp4nl)* status: 2. unresolved discussions.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45752)* status: :shipit: complete! ✓

**Comment by Abhishek-CDC** [comment]
trigger ut7

**Comment by Abhishek-CDC** [comment]
trigger ut0

**Comment by Abhishek-CDC** [comment]
trigger ut2

**Pull request merged**